x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/SourceOnlySnapshotRepository.java

```diff
@@ -0,0 +1,151 @@
+                    delegateType, metaData.settings()), typeLookup));
+            }
+        };
+    }
+
+    SourceOnlySnapshotRepository(Repository in) {
+        super(in);
+    }
+
+    @Override
+    public IndexMetaData getSnapshotIndexMetaData(SnapshotId snapshotId, IndexId index) throws IOException {
+        IndexMetaData snapshotIndexMetaData = super.getSnapshotIndexMetaData(snapshotId, index);
+        // TODO: can we lie about the index.version.created here and produce an index with a new version since we reindex anyway?
+
+        // for a minimal restore we basically disable indexing on all fields and only create an index
+        // that is fully functional from an operational perspective. ie. it will have all metadata fields like version/
+        // seqID etc. and an indexed ID field such that we can potentially perform updates on them or delete documents.
+        ImmutableOpenMap<String, MappingMetaData> mappings = snapshotIndexMetaData.getMappings();
+        Iterator<ObjectObjectCursor<String, MappingMetaData>> iterator = mappings.iterator();
+        IndexMetaData.Builder builder = IndexMetaData.builder(snapshotIndexMetaData);
+        while (iterator.hasNext()) {
+            ObjectObjectCursor<String, MappingMetaData> next = iterator.next();
+            MappingMetaData.Routing routing = next.value.routing();
+            final String mapping;
+            if (routing.required()) { // we have to respect the routing to be on the safe side so we pass this one on.
+                mapping = "{ \"" + next.key + "\": { \"enabled\": false, \"_meta\": " + next.value.source().string() + ", " +
+                    "\"_routing\" : { \"required\" : true } } }";
+            } else {
+                mapping = "{ \"" + next.key + "\": { \"enabled\": false, \"_meta\": " + next.value.source().string() + " } }";
+            }
+            builder.putMapping(next.key, mapping);
+        }
+        builder.settings(Settings.builder().put(snapshotIndexMetaData.getSettings())
+            .put(SOURCE_ONLY.getKey(), true)
+            .put("index.blocks.write", true)); // read-only!
+        return builder.build();
+    }
+
+    @Override
+    public void snapshotShard(IndexShard shard, Store store, SnapshotId snapshotId, IndexId indexId, IndexCommit snapshotIndexCommit,
+                              IndexShardSnapshotStatus snapshotStatus) {
+        if (shard.mapperService().documentMapper() != null // if there is no mapping this is null
+            && shard.mapperService().documentMapper().sourceMapper().isComplete() == false) {
+            throw new IllegalStateException("Can't snapshot _source only on an index that has incomplete source ie. has _source disabled " +
+                "or filters the source");
+        }
+        ShardPath shardPath = shard.shardPath();
+        Path dataPath = shardPath.getDataPath();
+        // TODO should we have a snapshot tmp directory per shard that is maintained by the system?
```

Comment:
can you clarify what you mean?

Reply:
yeah so today I just create a directory which is private to this usecase. I wonder if we should do that or have a tmp directory per shard that plugins can use for stuff like this.